cpp/src/arrow/compute/registry.cc

```diff
@@ -115,20 +181,47 @@ std::unique_ptr<FunctionRegistry> FunctionRegistry::Make() {
   return std::unique_ptr<FunctionRegistry>(new FunctionRegistry());
 }
 
-FunctionRegistry::FunctionRegistry() { impl_.reset(new FunctionRegistryImpl()); }
+std::unique_ptr<FunctionRegistry> FunctionRegistry::Make(FunctionRegistry* parent) {
+  return std::unique_ptr<FunctionRegistry>(
+      new FunctionRegistry(new FunctionRegistry::FunctionRegistryImpl(&*parent->impl_)));
```

Comment:
```suggestion
      new FunctionRegistry(parent->impl_.get());
```
The way it is currently will create a copy of the parent.  This pointer leaks but, more importantly, if you did something like...

```
auto parent = GetFunctionRegistry();
auto nested = FunctionRegistry::Make(parent);
// The function added here (using parent) will not be visible by nested
// and, if I'm understanding the purpose correctly, we would want that
parent->AddFunction(...);
```

Instead we should just take a non-owning "view" pointer to the parent.  If the parent is deleted before the nested registry is deleted then bad things will happen but that is to be expected.